src/usb.c

```diff
@@ -0,0 +1,1619 @@
+                                             "Not allowed to access USB devices");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  if (sender_info->sender_state != USB_SENDER_STATE_ACQUIRING_DEVICES || sender_info->acquiring_devices == NULL)
+    {
+      /* If the request was cancelled in some way. This would happen by calling
+       * FinishAcquireDevices after the user denied the permission.
+       */
+      sender_info->sender_state = USB_SENDER_STATE_DEFAULT;
+      g_clear_pointer (&sender_info->acquiring_devices, g_ptr_array_unref);
+
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                             "There is no device being acquired");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  /* We should never trigger these asserts. */
+  g_assert (sender_info->sender_state == USB_SENDER_STATE_ACQUIRING_DEVICES);
+  g_assert (sender_info->acquiring_devices != NULL);
+
+  fds = g_unix_fd_list_new ();
+
+  g_variant_builder_init (&results_builder, G_VARIANT_TYPE ("a(sa{sv})"));
+
+  accessed_devices = 0;
+  while (accessed_devices < MAX_DEVICES &&
+         sender_info->acquiring_devices->len > 0)
+    {
+      g_autoptr(UsbDeviceAcquireData) access_data = NULL;
+      g_autoptr(GError) error = NULL;
+      xdp_autofd int fd = -1;
+      GVariantDict dict;
+      GUdevDevice *device;
+      UsbOwnedDevice *owned_device = NULL;
+      const char *device_file;
+      int fd_index;
+
+      g_variant_dict_init (&dict, NULL);
+
+      access_data = g_ptr_array_steal_index (sender_info->acquiring_devices, 0);
+      device = g_hash_table_lookup (self->ids_to_devices, access_data->device_id);
+
+      if (!device)
+        {
+          g_variant_dict_insert (&dict, "success", "b", FALSE);
+          g_variant_dict_insert (&dict, "error", "s", _("Device not available"));
+          g_variant_builder_add (&results_builder, "(s@a{sv})",
+                                 access_data->device_id,
+                                 g_variant_dict_end (&dict));
+          continue;
+        }
+
+      device_file = g_udev_device_get_device_file (device);
+      g_assert (device_file != NULL);
+
+      /* Can the app even request this device? */
+      if (!usb_sender_info_match_device (sender_info, device))
+        {
+          g_variant_dict_insert (&dict, "success", "b", FALSE);
+          g_variant_dict_insert (&dict, "error", "s", _("Not allowed"));
+          g_variant_builder_add (&results_builder, "(s@a{sv})",
+                                 access_data->device_id,
+                                 g_variant_dict_end (&dict));
+          continue;
+        }
+
+      /* Check we haven't already acquired the device */
+      owned_device = g_hash_table_lookup (sender_info->owned_devices, access_data->device_id);
+      if (!owned_device)
```

Comment:
Why not first ensure that there is an owned device and then add it to the vardict and to the hashtable?

Then you could also move the code to actually open the device to the OwnedDevice